### PR TITLE
Interceptor for Java EE 

### PR DESCRIPTION
**Pull request by normalian**
This is Java EE Interceptor patch for AppInsights-Java. To enable this for applications, we have to setup two steps like below.

##  Create beans.xml file
to enable interceptor, we have to locate beans.xml ( e.g WEB-INF/beans.xml ) like below.

```<?xml version="1.0" encoding="UTF-8"?>
<beans xmlns="http://java.sun.com/xml/ns/javaee" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://java.sun.com/xml/ns/javaee http://java.sun.com/xml/ns/javaee/beans_1_0.xsd">
	<interceptors>
		<class>com.microsoft.applicationinsights.web.javaee.RequestNameInterceptor</class>
	</interceptors>
</beans>```

##  Add RequestName annotation to classes
Add RequestName annotation to monitored classes like below
```
@RequestScoped
@Named
public class IndexAction {

    @RequestName
    public String do1() {
        System.out.println("IndexAction#do1() =");
        return "0";
    }
}
```

### CODE REVIEW

File: web/src/main/java/com/microsoft/applicationinsights/web/javaee/RequestName.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
+/*
```
Comment (amnonshmsft) [inline web/src/main/java/com/microsoft/applicationinsights/web/javaee/RequestName.java]: Please add the license statement as in RequestNameInterceptorTests.java

File: web/src/main/java/com/microsoft/applicationinsights/web/javaee/RequestNameInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
+/*
```
Comment (amnonshmsft) [inline web/src/main/java/com/microsoft/applicationinsights/web/javaee/RequestNameInterceptor.java]: Please add the license statement as in RequestNameInterceptorTests.java

File: web/build.gradle [hunk 3]
```diff
@@ -29,6 +29,7 @@ dependencies {
     compile 'com.opensymphony:xwork:2.0.4' // Struts 2
     compile 'org.springframework:spring-webmvc:3.1.0.RELEASE'
     compile group: 'javax.servlet', name: 'javax.servlet-api', version: '3.0.1'
+	compile group: 'javax.enterprise', name: 'cdi-api', version: '1.1' // Java EE
```
Comment (amnonshmsft) [inline web/build.gradle]: Please use spaces (vs. tabs)

**Comment by amnonshmsft** [comment]
**normalian** - thanks a lot for your contribution. 

Before accepting an external pull requests we need the requester to execute a contributor license agreement (CLA) that secures adequate rights for Microsoft to redistribute your contributions. This can be done via **https://cla.microsoft.com/**.

Thanks!

File: web/src/main/java/com/microsoft/applicationinsights/web/javaee/RequestNameInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        setRequestNameSafe(ic);
+
+        return ic.proceed();
+    }
+
+    private void setRequestNameSafe(InvocationContext ic) {
+        try {
+            RequestTelemetryContext context = ThreadContext.getRequestTelemetryContext();
+            if (context != null) {
+
```
Comment (amnonshmsft) [inline web/src/main/java/com/microsoft/applicationinsights/web/javaee/RequestNameInterceptor.java]: Please remove redundant line

File: web/src/test/java/com/microsoft/applicationinsights/web/javaee/RequestNameInterceptorTests.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        RequestTelemetryContext requestTelemetryContext = new RequestTelemetryContext(DateTimeUtils.getDateTimeNow().getTime());
+        ThreadContext.setRequestTelemetryContext(requestTelemetryContext);
+        RequestTelemetry requestTelemetry = ThreadContext.getRequestTelemetryContext().getHttpRequestTelemetry();
+        requestTelemetry.setHttpMethod(HTTP_METHOD);
+
+        // Setting mock for action context
+        // We can't stub classes of Method and Class, so We should use RequestNameInterceptorTests class for stub
+        clazz = this.getClass();
+        method = clazz.getMethod("classInitialize");
+    	invocationContext = mock(InvocationContext.class);
```
Comment (amnonshmsft) [inline web/src/test/java/com/microsoft/applicationinsights/web/javaee/RequestNameInterceptorTests.java]: Please add another space before `invocationContext = mock(InvocationContext.class);`

File: web/src/test/java/com/microsoft/applicationinsights/web/javaee/RequestNameInterceptorTests.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    public void testInterceptorSetRequestNameCorrectly() throws Exception {
+        interceptor.invoke(invocationContext);
+
+        RequestTelemetry requestTelemetry = ThreadContext.getRequestTelemetryContext().getHttpRequestTelemetry();
+        Assert.assertEquals(REQUEST_NAME, requestTelemetry.getName());
+    }
+
+    @Test
+    public void testActionInvocationWhenExceptionThrownDuringRequestCalculation() throws Exception {
+
```
Comment (amnonshmsft) [inline web/src/test/java/com/microsoft/applicationinsights/web/javaee/RequestNameInterceptorTests.java]: In all files (and specifically in this test) please replace tabs with 4 spaces

**Comment by normalian** [comment]
Thank you for your messaging.

I have submit the CLA, please check it.

 

Regards

Daichi Isami

**Comment by amnonshmsft** [comment]
Daichi – thanks!
I’ll look into it soon.
In the meanwhile please take a look at the pull request comments and fix.
Thanks again,
Amnon

**Pull request merged**